compiler/rustc_const_eval/src/transform/validate.rs

```diff
@@ -285,12 +294,41 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                     };
                     check_equal(self, location, f_ty);
                 }
-                ty::Generator(_, substs, _) => {
-                    let substs = substs.as_generator();
-                    let Some(f_ty) = substs.upvar_tys().nth(f.as_usize()) else {
-                        fail_out_of_bounds(self, location);
-                        return;
+                ty::Generator(def_id, substs, _) => {
+                    let f_ty = if let Some(var) = parent_ty.variant_index {
+                        let gen_body = if *def_id == self.body.source.def_id() {
+                            self.body
+                        } else {
+                            self.tcx.optimized_mir(def_id)
+                        };
+
+                        let Some(layout) = gen_body.generator_layout() else {
+                            self.fail(location, format!("No generator layout for {:?}", parent_ty));
+                            return;
+                        };
+
+                        let Some(&local) = layout.variant_fields[var].get(f) else {
+                            fail_out_of_bounds(self, location);
+                            return;
+                        };
+
+                        let Some(&f_ty) = layout.field_tys.get(local) else {
+                            self.fail(location, format!("Out of bounds local {:?} for {:?}", local, parent_ty));
+                            return;
+                        };
+
+                        f_ty
+                    } else {
+                        let substs = substs.as_closure();
+
+                        let Some(f_ty) = substs.upvar_tys().nth(f.as_usize()) else {
+                            fail_out_of_bounds(self, location);
+                            return;
+                        };
+
+                        f_ty
```

Comment:
is that branch reachable? I guess more importantly, should that branch be reachable?

Reply:
The [last failure](https://github.com/rust-lang/rust/runs/6363521456?check_suite_focus=true) was because the variant index was None, so it is reachable. I have no idea if it should be though.